### PR TITLE
confirmation dialog before adding card type

### PR DESCRIPTION
**Pull request by Akshay0701**
cleared todo: added confirmation dialog before adding cards, same as Anki Desktop

## Anki Desktop
Capture

## AnkiDroid
Capture

### CODE REVIEW

**Comment by david-allison-1** [comment]
Hi, unit tests are failing. And could you add a screenshot?

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
unit tests need attention

**Comment by Akshay0701** [comment]
Yes i will fix it

**Comment by Akshay0701** [comment]
David, I tried to fix it.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java [hunk 1]
```diff
@@ -188,6 +188,17 @@ public void testTemplateAdd() throws Exception {
...
+        int numAffectedCards = 0;
+        if (!TemporaryModel.isOrdinalPendingAdd(testEditor.getTempModel(), ordinal)) {
+            Timber.d("Ordinal is not a pending add, so we'll get the current card count for confirmation");
+            numAffectedCards = getCol().getModels().tmplUseCount(testEditor.getTempModel().getModel(), ordinal);
+        }
+        Assert.assertEquals("Did not show dialog about adding template and it's card?",
+                getQuantityString(R.plurals.card_template_editor_confirm_add, numAffectedCards, numAffectedCards),
+                getDialogText(true));
+        clickDialogButton(DialogAction.POSITIVE, true);
+        advanceRobolectricLooperWithSleep();
```
Comment (Akshay0701) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java:201]: These adding and deleting cards lines are repeated too many times, can I extract it to methods?

Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.java:201]: Please, feel free to do any code improvement you think may benefit the codebase. Exctracting repeated code is almost always a good idea

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 2]
```diff
@@ -631,9 +631,14 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
             if (itemId == R.id.action_add) {
                 Timber.i("CardTemplateEditor:: Add template button pressed");
-                // TODO in Anki Desktop, they have a popup first with "This will create %d cards. Proceed?"
-                //      AnkiDroid never had this so it isn't a regression but it is a miss for AnkiDesktop parity
-                addNewTemplateWithCheck(tempModel.getModel());
+                // Show confirmation dialog
+                int ordinal = mTemplateEditor.mViewPager.getCurrentItem();
+                int numAffectedCards = 0;
+                if (!TemporaryModel.isOrdinalPendingAdd(tempModel, ordinal)) {
+                    Timber.d("Ordinal is not a pending add, so we'll get the current card count for confirmation");
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:638]: I doubt it's relevant as a Timber. As it makes reference to "ordinal" without giving it, and require to know what a pending add is. There is either too much or too little information. It looks mostly like a comment here.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java [hunk 3]
```diff
@@ -631,9 +631,14 @@ public boolean onOptionsItemSelected(MenuItem item) {
             int itemId = item.getItemId();
             if (itemId == R.id.action_add) {
                 Timber.i("CardTemplateEditor:: Add template button pressed");
-                // TODO in Anki Desktop, they have a popup first with "This will create %d cards. Proceed?"
-                //      AnkiDroid never had this so it isn't a regression but it is a miss for AnkiDesktop parity
-                addNewTemplateWithCheck(tempModel.getModel());
+                // Show confirmation dialog
+                int ordinal = mTemplateEditor.mViewPager.getCurrentItem();
+                int numAffectedCards = 0;
+                if (!TemporaryModel.isOrdinalPendingAdd(tempModel, ordinal)) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:637]: Can you please indicate in which case the condition will be false. Make a comment repeating what is a pending add maybe. 
It seems that if this is false, you'll show a message warning that it'll add 0 cards, it seems useless.

**Review by Arthur-Milchior — COMMENTED** [review]
Can you please replace "adding cards" by "adding card type" everywhere (e.g. commit title), because you are adding card types. It is true that it adds card as a consequence, but "adding card" do not describe the full impact.
Furthermore, the window to add notes is called "add card" so it's even more confusing.


Does the test works with only your first commit? We need test to work on all commits, so if the second commit is required for test to work, please put all code change together

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java (hunk 3 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:637]: You're right!,
this if statement checking if there are any pending process of "add cards types", if it has pending process then we should not run `confirmAddCards`.
maybe I need to put `confirmAddCards` inside the if statement.

Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java:637]:
```
isOrdinalPendingAdd method will check if there are any new card types added or not,
if TempModel has new card type then numAffectedCards will be 0 by default.
```

**Comment by Akshay0701** [comment]
Done, squash commits into one and extracted method.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java (hunk 2 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.java]: yes, timber is not needed here, I added comments here.

**Review by mikehardy — APPROVED** [review]
@Akshay0701 you are a brave person, going in to the CardTemplateEditor. This all looks correct so I will hereby induct you in to the select group known as the Knights of the Ordinal, those who actually understand Anki's Notes/Fields/Template Ordinals/Cards relationship thoroughly ;-).
Also you're now responsible for any CardTemplateEditor bugs :laughing: 

Arthur had one issue which it appears you addressed, so I'm going to merge this as it all looks good to me and most of this mess of code was mine to begin with so it's one of the few areas I feel really confident on

Cheers

**Comment by Arthur-Milchior** [comment]
Would you mind indicating what mean the "string" label. I see only two PR have it, so I assume it's not all PR that add a new strings

**Comment by Arthur-Milchior** [comment]
@Akshay0701 would you mind taking a screen shot that is not trimmed?
One goal of the screenshot is to show the translators the context in which the string appear. With this trimmed  screenshot, they see it's a ok/cancel box, but they lack the card editor context

**Comment by Akshay0701** [comment]
@Arthur-Milchior is this okay ?

**Comment by Arthur-Milchior** [comment]
Uploaded on crowdin

**Comment by Arthur-Milchior** [comment]
yes, thanks

**Pull request merged**